src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs

```diff
@@ -25,83 +25,255 @@ internal sealed partial class TarHeader
+                }
+            }
+            else // seekable archive stream, unseekable data stream
             {
-                WriteData(archiveStream, _dataStream, _size);
+                WriteWithUnseekableStreamAs(format, archiveStream, buffer);
             }
         }
 
-        // Asynchronously writes the current header as a V7 entry into the archive stream and returns the value of the final checksum.
-        internal async Task WriteAsV7Async(Stream archiveStream, Memory<byte> buffer, CancellationToken cancellationToken)
+        internal async Task WriteAsAsync(TarEntryFormat format, Stream archiveStream, Memory<byte> buffer, CancellationToken cancellationToken)
         {
-            cancellationToken.ThrowIfCancellationRequested();
+            Debug.Assert(format > TarEntryFormat.Unknown && format <= TarEntryFormat.Gnu);
+            Debug.Assert(archiveStream.CanSeek || _dataStream == null || _dataStream.CanSeek);
 
-            WriteV7FieldsToBuffer(buffer.Span);
+            if (_dataStream == null || _dataStream.CanSeek) // seek status of archive does not matter
+            {
+                switch (format)
+                {
+                    case TarEntryFormat.V7:
+                        WriteV7FieldsToBuffer(GetTotalDataBytesToWrite(), buffer.Span);
+                        break;
+                    case TarEntryFormat.Ustar:
+                        WriteUstarFieldsToBuffer(GetTotalDataBytesToWrite(), buffer.Span);
+                        break;
+                    case TarEntryFormat.Pax:
+                        WritePaxFieldsToBuffer(GetTotalDataBytesToWrite(), buffer.Span);
+                        break;
+                    case TarEntryFormat.Gnu:
+                        WriteGnuFieldsToBuffer(GetTotalDataBytesToWrite(), buffer.Span);
+                        break;
+                }
 
-            await archiveStream.WriteAsync(buffer, cancellationToken).ConfigureAwait(false);
+                await archiveStream.WriteAsync(buffer, cancellationToken).ConfigureAwait(false);
 
-            if (_dataStream != null)
+                if (_dataStream != null)
+                {
+                    await WriteDataAsync(archiveStream, _dataStream, _size, cancellationToken).ConfigureAwait(false);
+                }
+            }
+            else // seekable archive stream, unseekable data stream
             {
-                await WriteDataAsync(archiveStream, _dataStream, _size, cancellationToken).ConfigureAwait(false);
+                await WriteWithUnseekableDataStreamAsAsync(format, archiveStream, buffer, cancellationToken).ConfigureAwait(false);
             }
         }
 
+        private void WriteWithUnseekableStreamAs(TarEntryFormat format, Stream archiveStream, Span<byte> buffer)
+        {
+            // When the data stream is unseekable, the order in which we write the entry data changes
+            Debug.Assert(archiveStream.CanSeek);
+            Debug.Assert(_dataStream != null);
```

Comment:
Good point. Although I prefer to put it in its own separate debug assert, so that when/if it fails, it's quick and easy to determine exactly which condition failed.